### PR TITLE
Facet expand

### PR DESCRIPTION
**Pull request by Fil**
Expand all channels in parallel to the facet reindexation.

~~Alternatives: #1068, #1070.~~

TODO:
- [x] create a central registry of channels, with their definitions (from [facets.js#L56](https://github.com/observablehq/plot/pull/1069/files#diff-540ac050c947f485398ca75ea672a4d4a23c59e708346f8a5dda03408b195ab7R56)) and associated scales (from [plot.js#L756](https://github.com/observablehq/plot/blob/f7aa3b23bcadd4911c216b1a82ee5618c92b27c9/src/plot.js#L756)).
- [x] put the "other" loop upstream
- ~~[x] offer an option to expand more (unlisted) channels—or to list more channels(?)~~ we'll expose registerChannel later

### CODE REVIEW

File: src/channel.js [hunk 1]
```diff
@@ -2,6 +2,53 @@ import {ascending, descending, rollup, sort} from "d3";
...
 import {registry} from "./scales/index.js";
 import {maybeReduce} from "./transforms/group.js";
+import {maybeColorChannel, maybeNumberChannel} from "./options.js";
+import {maybeSymbolChannel} from "./symbols.js";
+import {warn} from "./warnings.js";
+
+// An array of known channels, with an associated scale name, and a definition
+// that returns [variable, undefined] if variable, or [undefined, constant] if
+// constant (such as "#eee" for the color channel)
+export const knownChannels = new Map([
```
Comment (mbostock) [inline src/channel.js:12]: Let’s call this `registry` (or `channelRegistry`) like we do for src/scales/index.js, and be consistent.

File: src/channel.js [hunk 2]
```diff
@@ -2,6 +2,53 @@ import {ascending, descending, rollup, sort} from "d3";
...
+  ["length", {scale: "length", definition: maybeNumberChannel}], // vector
+  ["width", {definition: maybeNumberChannel}], // image
+  ["height", {definition: maybeNumberChannel}], // image
+  ["src", {definition: maybePathChannel}], // image
+  ["weight", {definition: maybeNumberChannel}] // density
+]);
+
+export function registerChannel(name, {scale, definition} = {}) {
+  knownChannels.set(name, {scale, definition});
+}
```
Comment (mbostock) [inline src/channel.js:43]: Let’s leave this out for now. We don’t support this for scales yet either.

File: src/channel.js [hunk 3]
```diff
@@ -2,6 +2,53 @@ import {ascending, descending, rollup, sort} from "d3";
...
+
+export function definition(name, value, defaultValue) {
+  if (!knownChannels.has(name)) {
+    warn(`The ${name} channel is not registered and might be incompatible with some transforms.`);
+  }
+  const {definition} = knownChannels.get(name) || {};
+  return definition ? definition(value, defaultValue) : value === undefined ? [undefined, defaultValue] : [value];
```
Comment (mbostock) [inline src/channel.js:50]:
```suggestion
  return definition !== undefined ? definition(value, defaultValue)
    : value === undefined ? [undefined, defaultValue]
    : [value];
```

File: src/marks/cell.js [hunk 4]
```diff
@@ -54,7 +55,7 @@ export function cell(data, options = {}) {
  */
 export function cellX(data, options = {}) {
   let {x = indexOf, fill, stroke, ...remainingOptions} = options;
-  if (fill === undefined && maybeColorChannel(stroke)[0] === undefined) fill = identity;
+  if (fill === undefined && definition("stroke", stroke)[0] === undefined) fill = identity;
```
Comment (mbostock) [inline src/marks/cell.js:58]: Maybe maybeChannel would work as a more specific name here?

```suggestion
  if (fill === undefined && maybeChannel("stroke", stroke)[0] === undefined) fill = identity;
```

File: src/plot.js [hunk 5]
```diff
@@ -674,11 +675,14 @@ export class Mark {
...
-        throw new Error(`missing channel value: ${name}`);
-      })
+      Object.entries(channels)
+        .filter(([name, {value, optional}]) => {
+          if (!knownChannels.has(name)) warn(`The ${name} channel is not registered.`);
+          if (value != null) return true;
+          if (optional) return false;
+          throw new Error(`missing channel value: ${name}`);
+        })
+        .map(([name, {scale = (knownChannels.get(name) || {}).scale, ...rest}]) => [name, {scale, ...rest}])
```
Comment (mbostock) [inline src/plot.js:685]:
```suggestion
        .map(([name, {scale = knownChannels.get(name)?.scale, ...rest}]) => [name, {scale, ...rest}])
```

File: src/plot.js [hunk 6]
```diff
@@ -756,24 +760,8 @@ function applyScaleTransforms(channels, options) {
...
-          scale = "opacity";
-          break;
-        default:
-          scale = scaleRegistry.has(name) ? name : null;
-          break;
-      }
-      channel.scale = scale;
-    }
+    const {scale} = channel;
+    if (scale === true) channel.scale = knownChannels.get(name).scale;
```
Comment (mbostock) [inline src/plot.js:764]: 👏

Comment (mbostock) [reply src/plot.js:764]: Though maybe we want to handle an unknown channel as before?

```suggestion
    if (scale === true) channel.scale = knownChannels.get(name)?.scale ?? null;
```

File: src/transforms/dodge.js [hunk 7]
```diff
@@ -3,7 +3,7 @@ import {finite, positive} from "../defined.js";
 import {identity, maybeNamed, number, valueof} from "../options.js";
 import {coerceNumbers} from "../scales.js";
 import {initializer} from "./basic.js";
-
+import {exclusiveFacets, maybeExpand} from "../facet.js";
```
Comment (mbostock) [inline src/transforms/dodge.js:6]:
```suggestion
import {exclusiveFacets, maybeExpand} from "../facet.js";

```

**Comment by Fil** [comment]
superseded by #1649

**Pull request closed without merging**